sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -449,54 +520,155 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
 
     @Override
     public void validate(PBegin input) {
-      checkNotNull(getProjectId(), "projectId");
-      checkNotNull(getQuery(), "query");
+      checkNotNull(getProjectId(), "projectId ValueProvider");
+
+      if (getProjectId().isAccessible()) {
+        checkNotNull(getProjectId().get(), "projectId");
+      }
+
+      if (getQuery() == null && getGqlQuery() == null) {
+        throw new IllegalArgumentException(
+            "Either query or gql query ValueProvider should be provided");
+      }
+
+      if (getQuery() != null && getGqlQuery() != null) {
+        throw new IllegalArgumentException(
+            "Only one of query or gql query ValueProvider should be provided");
```

Comment:
My understanding here is that we do not want to allow specifying `Query` at runtime (might not be as useful, as the users will have to json-encode the Query). So somewhere the check has to be made to validate `query` and `gqlQuery`, if both or none are provided at construction time. 

I did consider a `getQueryValueProvider` method which uses a `NestedValueProvider` to either return a `query` or  translate the `gqlQuery` but did not see much benefit in doing so, and required more wiring given that the translator needs to talk to Datastore. Having the `GqlTranslatorFn` as separate `ParDo` in the graph makes it more explicit about the conversion. 
(getQuery is a AutoValue generated method, so cannot modify directly?)